### ISSUE TITLE
What is wrong with the ONNX Yolov5 model's inference result?

### ISSUE THREAD

**Issue by beveredge** [ultralytics/yolov5#2966]
## ❔Question
Use export.py to generate yolov5s.onnx for 640x640 dimension
Run inference for picture and got raw result from 3 layers of the network:

- (1, 3, 20, 20, 85)
- (1, 3, 40, 40, 85)
- (1, 3, 80, 80, 85)

Original:
![orig](https://user-images.githubusercontent.com/73838088/116427397-a8542600-a84c-11eb-84a9-f548aea98036.jpg)

I look through all the outputs and channels, trying to find the person in the picture

Result:
![frame1113](https://user-images.githubusercontent.com/73838088/116427601-d6d20100-a84c-11eb-9938-8c790ef590ed.jpg)

For output "417" have correct result. Raw data: result = [-0.0460704, 1.4253, -0.312967, -0.352123, 0.689956 ... ]

Center of object has 

- x = colNum + sigmoid(result[0])) * gridStepX
- y = rowNum + sigmoid(result[1])) * gridStepY

where colNum and rowNum - column and row numbers

Width and height of object has

- w = exp(result[2]) * anchor1 // anchor1 value is 59
- h = exp(result[3]) * anchor2 // anchor1 value is 119

Result box is correct: confidence 0.665087, x 119.816, y 428.899, w 43.1451, h 83.680

The same object for output "output" have incorrect result. Raw data: result = [0.857506, -0.272947, 0.230578, 2.81826, 0.223092, ... ]. Object has x 117.617, y 435.457, w 41.5578, but **h 385.197**

- w = exp(result[2]) * anchor1 // exp(0.230578) * 33 
- h = exp(result[3]) * ancho2 // exp(2.81826) * 23

The object has considerable confidence but incorrect height. What is wrong with result? How is it possible to ignore similar results?

**Comment by beveredge** [comment]
I use OpenCV DNN NMSBoxes. The result is unsatisfactory
![frame1113](https://user-images.githubusercontent.com/73838088/116434150-d50b3c00-a852-11eb-97fc-d321967e30cf.jpg)

**Comment by glenn-jocher** [comment]
@beveredge before exporting you should verify PyTorch inference, then simply run:
```
python models/export.py --grid --simplify
```

**Comment by beveredge** [comment]
ONNX: simplifier failure: [ONNXRuntimeError] : 1 : FAIL : Node (Mul_925) Op (Mul) [ShapeInferenceError] Incompatible dimensions
ONNX: export success, saved as ./yolov5s.onnx (29.2 MB)
CoreML: starting export with coremltools 4.1...
CoreML: export failure: 'torch._C.Node' object has no attribute 'ival'

Export complete (8.68s). Visualize with https://github.com/lutzroeder/netron.

**Comment by beveredge** [comment]
Problem was with box calculating. I used formulas for YoloV3 like in [comment](https://github.com/ultralytics/yolov5/issues/343#issuecomment-662317958). Must use for YoloV5

**Comment by hzitoun** [comment]
@beveredge Could you please provide the code you used to parse CV2 DNN ONNX output?

**Comment by beveredge** [comment]
You have 3 arrays with inference results: [1,3,80,80,85], [1,3,40,40,85] and [1,3,20,20,85]. Depending on the size, you must set the value of `anchors`. This value can be taken from the YOLOv3 network:
```
const unsigned LAYER_20_20_ANCH_Y3[3][2] = { {116, 90}, {156, 198}, {373, 326} };
const unsigned LAYER_40_40_ANCH_Y3[3][2] = { {30, 61},  {62, 45},   {59, 119} };
const unsigned LAYER_80_80_ANCH_Y3[3][2] = { {10, 13},  {16, 30},   {33, 23} };
```
Secondary functions:
```
//SoftMax transformation: converting data to an interval from 0 to 1
std::vector<float> softMax(const float* data, const int size)
{
    std::vector<float> buf;
    std::transform(data, data + size, back_inserter(buf), [](float item) { return std::exp(item); });
    float sum = std::accumulate(buf.begin(), buf.end(), 0.0);
    std::transform(buf.begin(), buf.end(), buf.begin(), [sum](float& item) { return item / sum; });
    return buf;
}

//Sigmoid value
float sigmoid(float value)
{
    return 1.0f / (1.0f + std::exp(-value));
}
```

You have the transformed image for inference `inferImage` and its original size `inputSize`. You also get the result of inference at the outputs of the neural network with the size [1,3,80,80,85], [1,3,40,40,85] or [1,3,20,20,85] with float type.

```
// Iterator for raw data
const float* resultIter = inferResult;

static const unsigned CONFIDENCE_INDEX {4};
static const unsigned CONF_DISTRIB_INDEX {5};
static const unsigned CHANNEL_STEP {85};
static const float CONFIDENCE_THRESHOLD {0.2};

// Calculate scales and grid steps
const float wSaleFactorToOrigin = float(inferImage.cols) / inputSize.width;
const float hSaleFactorToOrigin = float(inferImage.rows) / inputSize.height;
const int gridStepX = inputSize.width / shapeOutColumns;
const int gridStepY = inputSize.height / shapeOutRows;
```
For example: you got inference result [1,3,80,80,85], so shapeOutChannels == 1, shapeOutRows == 80, shapeOutColumns == 80
You must parse an array of 85 elements:
```
for (int channel{0}; channel != shapeOutChannels; ++channel)
{
    for (int row{0}; row != shapeOutRows; ++row)
    {
        for (int col{0}; col != shapeOutColumns; ++col)
        {
            // Read object confidence
            const float objConfid = sigmoid(resultIter[CONFIDENCE_INDEX]);

            // Skip object if confidence < CONFIDENCE_THRESHOLD
            if (objConfid < CONFIDENCE_THRESHOLD)
            {
                resultIter += CHANNEL_STEP;
                continue;
            }

            // Iterator for coco-features
            const float* confidBegin = (resultIter + CONF_DISTRIB_INDEX);
            
            // Take confidence distribution for coco-objects
            auto confDitrib = this->softMax(confidBegin, (CHANNEL_STEP - CONF_DISTRIB_INDEX));

            // Find position with max confidence
            const auto classConfidPos = std::max_element(confDitrib.begin(), confDitrib.end());
            
            // Take class confidence value
            const float classConfid = *classConfidPos;

            // Skip object if class confidence < CONFIDENCE_THRESHOLD
            if (classConfid < CONFIDENCE_THRESHOLD)
            {
                resultIter += CHANNEL_STEP;
                continue;
            }

            // Correct object confidence
            const float totalConfid = objConfid * classConfid;
            // Calculate coco object ID
            const auto classId = std::distance(confDitrib.begin(), classConfidPos);

            // Calculate box size
            float rawX = resultIter[0];
            float rawY = resultIter[1];
            float rawWidth = resultIter[2];
            float rawHeight = resultIter[3];

            float w = std::pow(2 * sigmoid(rawWidth), 2) * anchors[channel][0];
            float h = std::pow(2 * sigmoid(rawHeight), 2) * anchors[channel][1];
            float x = ((col + 2 * sigmoid(rawX) - 0.5) * gridStepX);
            float y = ((row + 2 * sigmoid(rawY) - 0.5) * gridStepY);

            //Result
            w *= wSaleFactorToOrigin;
            h *= hSaleFactorToOrigin;
            x *= wSaleFactorToOrigin;
            y *= hSaleFactorToOrigin;

            resultIter += CHANNEL_STEP;
        }
    }
} 
```

**Comment by guyiyifeurach** [comment]
> ONNX: simplifier failure: [ONNXRuntimeError] : 1 : FAIL : Node (Mul_925) Op (Mul) [ShapeInferenceError] Incompatible dimensions ONNX: export success, saved as ./yolov5s.onnx (29.2 MB) CoreML: starting export with coremltools 4.1... CoreML: export failure: 'torch._C.Node' object has no attribute 'ival'
> 
> Export complete (8.68s). Visualize with https://github.com/lutzroeder/netron.

Hi, can you tell me how to solve the first problem? I met the same problem, but I haven't solved it.
I use pt model to inference normally, and I export pt model to onnx model, it changed normally, but when I use onnx model to inference, I met the problem.

**Comment by glenn-jocher** [comment]
@guyiyifeurach ONNX export and inference is pretty simple:

```
python export.py --weights yolov5s.pt
python detect.py --weights yolov5s.onnx
```
